release.py

```diff
@@ -20,14 +20,110 @@
 import sys
 import tempfile
 from contextlib import contextmanager
-from typing import Any, Callable, Generator, Self
+from dataclasses import dataclass
+from pathlib import Path
+from typing import (
+    Any,
+    Callable,
+    Generator,
+    Literal,
+    Protocol,
+    Self,
+    overload,
+)
 
 COMMASPACE = ", "
 SPACE = " "
 tag_cre = re.compile(r"(\d+)(?:\.(\d+)(?:\.(\d+))?)?(?:([ab]|rc)(\d+))?$")
 
 
-# Ideas stolen from Mailman's release script, Lib/tokens.py and welease
+class ReleaseShelf(Protocol):  # this will be easier to type in 3.13
```

Comment:
This is like a `TypedDict` but for a pseudo-mapping not inheriting from a `dict`. The syntax is a bit horrible but it's per-key strongly typed.

And actually, this *won't* be easier in 3.13 :sob:
https://peps.python.org/pep-0705/#a-typedmapping-protocol-type